packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -108,64 +118,122 @@ export function memoizeTracked<T>(
+
+      cache = debugCache;
+    }
+
+    let { tag, snapshot, fn } = cache;
+
+    if (tag === undefined || !validateTag(tag, snapshot)) {
       beginTrackFrame();
 
       try {
         if (DEBUG) {
-          runInAutotrackingTransaction!(() => (lastValue = callback(...args)), debuggingContext);
+          runInAutotrackingTransaction!(() => (cache.lastValue = fn(...args)), debuggingContext!);
         } else {
-          lastValue = callback(...args);
+          cache.lastValue = fn(...args);
         }
       } finally {
         tag = endTrackFrame();
-        snapshot = valueForTag(tag);
+        cache.tag = tag;
+        cache.snapshot = valueForTag(tag);
         consumeTag(tag);
-
-        // If the final tag is constant, then we know for sure that this
-        // memoized function can never change. There are times when this
-        // information is useful externally (i.e. in the append VM, it tells us
-        // whether or not to emit opcodes) so we expose it via a metadata weakmap.
-        if (tag === CONSTANT_TAG) {
-          IS_CONST_MAP.set(memoized, true);
-        } else if (DEBUG) {
-          // In DEBUG, set the value to false explicitly. This way we can throw
-          // if someone attempts to call `isConst(memoized)` before running
-          // `memoized()` at least once.
-          IS_CONST_MAP.set(memoized, false);
-        }
       }
     } else {
       consumeTag(tag);
     }
 
-    return lastValue!;
-  };
-
-  if (DEBUG) {
-    IS_CONST_MAP.set(memoized, undefined);
+    return cache.lastValue!;
   }
 
-  return memoized;
-}
+  static isConst(cache: Cache) {
+    if (DEBUG) {
+      let debugCache = (cache as any)[DEBUG_CACHE];
+
+      if (!(debugCache instanceof Cache)) {
+        throw new Error(
+          `isConst() can only be used on an instance of a cache created with createCache(). Called with: ${String(
+            cache
+          )}`
+        );
+      }
 
-export function isConstMemo(memoized: Function) {
-  if (DEBUG && IS_CONST_MAP.has(memoized) && IS_CONST_MAP.get(memoized) === undefined) {
-    throw new Error(
-      'Attempted to call `isConstMemo` on a memoized function, but the function has not been run at least once yet. You cannot know if a memoized function is constant or not until it has been run at least once. Call the function, then pass it to `isConstMemo`.'
-    );
+      if (debugCache.tag === undefined) {
+        throw new Error(
+          `isConst() can only be used on a cache once getValue() has been called at least once. Called with cache function:\n\n${String(
+            debugCache.fn
+          )}`
+        );
+      }
+
+      cache = debugCache;
+    }
+
+    return isConstTag(cache.tag!);
   }
+}
 
-  return IS_CONST_MAP.get(memoized) === true;
+export function createCache<T>(fn: () => T): Cache<T> {
+  return new Cache(fn);
 }
 
+export const getValue = Cache.getValue;
+
+export const isConst = Cache.isConst;
```

Comment:
I think the code would be much easier to understand (and less error prone) if the implementation were moved into these functions (and out of the static functions).

Reply:
I really prefer using `static` methods over Symbols, but we can convert to symbols if you prefer. There were perf concerns when we were writing MonomorphicTagImpl, if I recall, I think we saw that using Symbols did have some amount of impact, but it was minor.

Reply:
FWIW, this technique also works with true private fields and methods, so I think it would be valuable to use. We could for instance remove the need for the DEBUG logic and wrapper function that way in the future. Or maybe even now if transpilers handle it well enough...

Reply:
I don't understand why the question is symbol vs static method. The static methods don't even use `this` (and are completely disconnected from the `this` binding).